File: binary_search_step.java
import java.io.*;
import java.util.*;

public class binary_search_step {
    private final List<String> steps = new ArrayList<>();

    public void recordStep(String step) {
        steps.add(step);
    }

    public void writeSteps(String filePath) throws IOException {
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(filePath))) {
            for (String step : steps) {
                bw.write(step);
                bw.newLine();
            }
        }
    }

    public List<String> getSteps() {
        return steps;
    }

    public static void main(String[] args) throws IOException {
        // List all .csv files in /Java
        File datasetDir = new File("Java");
        File[] csvFiles = datasetDir.listFiles((dir, name) -> name.endsWith(".csv"));
        if (csvFiles == null || csvFiles.length == 0) {
            System.out.println("No .csv files found.");
            return;
        }
        System.out.println("Available CSV files in /Java:");
        for (int i = 0; i < csvFiles.length; i++) {
            System.out.printf("%d: %s\n", i + 1, csvFiles[i].getName());
        }
        int fileChoice = -1;
        Scanner scanner = new Scanner(System.in);
        while (fileChoice < 1 || fileChoice > csvFiles.length) {
            System.out.print("Select a file by number: ");
            try {
                fileChoice = Integer.parseInt(scanner.nextLine().trim());
            } catch (NumberFormatException e) {
                fileChoice = -1;
            }
        }
        String filename = csvFiles[fileChoice - 1].getPath();

        System.out.print("Enter target number to search: ");
        int target = Integer.parseInt(scanner.nextLine().trim());

        List<Integer> numData = new ArrayList<>();
        // Read file to put data
        try (BufferedReader readFile = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = readFile.readLine()) != null) {
                String[] parts = line.split(",", 2);
                if (parts.length >= 1) {
                    int num = Integer.parseInt(parts[0]);
                    numData.add(num);
                }
            }
        } catch(IOException e){
            System.out.println("Error! File not found");
            return;
        }
        int count = numData.size();

        binary_search_step stepRecorder = new binary_search_step();
        int low = 0, high = count - 1, foundIndex = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            // Record the step as row: value (1-based row number)
            stepRecorder.recordStep((mid + 1) + ": " + numData.get(mid));
            if (numData.get(mid) == target) {
                foundIndex = mid;
                break;
            } else if (numData.get(mid) < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        if (foundIndex == -1) {
            stepRecorder.recordStep("-1");
        }
        String stepsOutput = String.format("Java/binary_search_step_%d.txt", target);
        stepRecorder.writeSteps(stepsOutput);
        System.out.println("\n--------------------------------");
        System.out.printf("Binary search steps saved to %s\n", stepsOutput);
        if (foundIndex != -1) {
            System.out.println("Target found at row: " + (foundIndex + 1));
        } else {
            System.out.println("Target not found");
        }
        System.out.println("--------------------------------");
    }
} 

File: binary_search_step.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>

using namespace std;

const int MAX_SIZE = 1000; // adjust based on dataset size

int main() {
    string filename;
    int target;

    // Get user input
    cout << "Enter dataset filename: ";
    getline(cin, filename);

    cout << "Enter target integer: ";
    cin >> target;

    // Open CSV file
    ifstream infile(filename);
    if (!infile) {
        cerr << "Error: Cannot open file " << filename << "\n";
        return 1;
    }

    int nums[MAX_SIZE];
    string strs[MAX_SIZE];
    int size = 0;
    string line;

    // Read from CSV into arrays
    while (getline(infile, line) && size < MAX_SIZE) {
        if (!line.empty()) {
            stringstream ss(line);
            string numPart, strPart;
            getline(ss, numPart, ',');
            getline(ss, strPart);
            nums[size] = stoi(numPart);
            strs[size] = strPart;
            ++size;
        }
    }
    infile.close();

    // Output file setup
    string outFile = "binary_search_step_" + to_string(target) + ".txt";
    ofstream outfile(outFile);
    if (!outfile) {
        cerr << "Error: Cannot write to file " << outFile << "\n";
        return 1;
    }

    // Binary search
    int left = 0, right = size - 1;
    bool found = false;

    while (left <= right) {
        int mid = left + (right - left) / 2;
        outfile << (mid + 1) << ": " << nums[mid] << "/" << strs[mid] << "\n";

        if (nums[mid] == target) {
            found = true;
            break;
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    if (!found) {
        outfile << "-1\n";
    }

    cout << "Search complete. Output written to: " << outFile << "\n";
    return 0;
}
